File: dayOneP1.php
<?php
/**
 * @var string $input
 */

$input = trim($input);
$pairs = explode("\n", $input);


$first_list = [];
$second_list = [];

foreach ($pairs as $pair)
{
    [$one, $two] = explode("   ", $pair);

    // echo "one: " . json_encode($one) . "\n";
    // echo "two: " . json_encode($two) . "\n\n";

    $first_list[] = $one;
    $second_list[] = $two;
}

// echo "first_list: " . json_encode($first_list) . "\n";
// echo "second_list: " . json_encode($second_list) . "\n";

sort($first_list, SORT_NUMERIC);
sort($second_list, SORT_NUMERIC);

// echo "first_list(sort): " . json_encode($first_list) . "\n";
// echo "second_list(sort): " . json_encode($second_list) . "\n";

$sum = 0;

for ($i = 0; $i < sizeof($first_list); $i++)
{
    $first = (int)$first_list[$i];
    $second = (int)$second_list[$i];

    $diff = abs($first - $second);

    // echo "diff: " . $diff . "\n";

    $sum += $diff;
}

echo $sum;

File: dayOneP1.c
#include "lib.c"
#include <stdio.h>
#include <string.h>


int
pivot(int* nums, int l, int r)
{
    int p = nums[r];

    int i = l - 1;

    for (int j = l; j < r; j++) {
        if (nums[j] < p) {
            i++;
            int tmp = nums[j];
            nums[j] = nums[i];
            nums[i] = tmp;
        }
    }

    i++;
    nums[r] = nums[i];
    nums[i] = p;

    return i;
}

void
qs(int* nums, int l, int r)
{
    if (l >= r)
        return;

    int p = pivot(nums, l, r);

    qs(nums, l, p - 1);
    qs(nums, p + 1, r);
}


int
solve(char* input)
{
    size_t size = strlen(input);

    size_t num_pairs = 1; // one for the first line
    for (int i = 0; i < size; i++)
        if (input[i] == '\n') num_pairs++;

    int* first_list = malloc(num_pairs * sizeof(int));
    size_t first_list_size = 0;

    int* second_list = malloc(num_pairs * sizeof(int));
    size_t second_list_size = 0;

    for (int i = 0; i < size; i++) {
        int num = 0;

        for (;input[i] != ' '; i++) {
            num *= 10;
            num += (input[i] - '0');
        }

        first_list[first_list_size++] = num;

        // skip the three whitespaces
        i += 3;

        num = 0;
        for (;input[i] != '\n'; i++) {
            num *= 10;
            num += (input[i] - '0');
        }

        second_list[second_list_size++] = num;
    }

    qs(first_list, 0, first_list_size - 1);
    qs(second_list, 0, second_list_size - 1);

    int sum = 0;

    for (int i = 0; i < first_list_size; i++) {
        sum += abs(first_list[i] - second_list[i]);
    }

    free(first_list);
    free(second_list);

    return sum;
}

int
main()
{
    char* input = get_input_day("one");

    // printf("%s", input);
    int res = solve(input);
    printf("%d\n", res);

    free(input);
    return 0;
}
